### ISSUE TITLE
Allow additional parameters in `compute_mat`

### ISSUE THREAD

**Issue by celsofranssa** [KevinMusgrave/pytorch-metric-learning#697]
Hello,

The following code snippet shows a custom distance function that scales the simple dot distance with the rewards associated with each embedding.

```python
import torch
from pytorch_metric_learning.distances import BaseDistance


class CustomDistance(BaseDistance):

    def __init__(self, params):
        super().__init__(params, is_inverted=True)
        assert self.is_inverted
        # dict of ref_emb rewards
        self.rewards = params.rewards

    def compute_mat(self, embeddings_ids, embeddings, ref_emb_ids, ref_emb):
        mat =  20 * torch.einsum("ab,cb->ac", embeddings, ref_emb)
        for col, ref_emb_idx in ref_emb_ids.items():
            mat[:, col] *= self.rewards[ref_emb_idx]
````

However, the loss function call of `mat = self.distance(embeddings, ref_emb)` not allows calling the overridden method `compute_mat(self, embeddings_ids, embeddings, ref_emb_ids, ref_emb)` (containing the embeddings and ref_emb ids) required to get the embedding reward and scale the corresponding distance value.

Is there a workaround?

Thank you.

**Comment by celsofranssa** [comment]
Hello @KevinMusgrave 
Do you have any suggestions here?

**Comment by KevinMusgrave** [comment]
Or do you mean you need to change the `mat = self.distance(embeddings, ref_emb)` in the loss function?

**Comment by celsofranssa** [comment]
> Would overriding the `forward` method work?

Exactly @KevinMusgrave, 

I am already using the very nice [NTXentLoss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#ntxentloss). Recently, I discovered that scaling the simple dot distance with the rewards associated with each embedding is beneficial for my research task. At this point, I need to access the embedding identifier because each embedding has its associated reward.

**Comment by KevinMusgrave** [comment]
I see, so you also need to modify NTXentLoss to take in those ids?

**Comment by celsofranssa** [comment]
> I see, so you also need to modify NTXentLoss to take in those ids?

If there is no other overriding approach, yes.

**Comment by KevinMusgrave** [comment]
Yeah I don't think there's another approach. The only other way would be to set some attribute of your custom distance object before computing the loss.

```python
dist_fn = CustomDistance()
loss_fn = NTXentLoss(distance=dist_fn)

...
dist_fn.curr_ref_ids = ref_ids
loss = loss_fn(...)
```

**Comment by celsofranssa** [comment]
> Yeah I don't think there's another approach. The only other way would be to set some attribute of your custom distance object before computing the loss.
> 
> ```python
> dist_fn = CustomDistance()
> loss_fn = NTXentLoss(distance=dist_fn)
> 
> ...
> dist_fn.curr_ref_ids = ref_ids
> loss = loss_fn(...)
> 
> # inside dist_fn refer to self.curr_ref_ids
> ```

I see. For a feature release, it would be great if the `mat = self.distance(embeddings, ref_emb)` function accepts additional parameters to leverage some custom distance implementations.

**Comment by celsofranssa** [comment]
Hello, 

Is there any progress here?

**Comment by KevinMusgrave** [comment]
Sorry, no progress yet
